### DISCUSSION TITLE
Report Outdated Components (OWASP Top 10: A06:2021)

### DISCUSSION THREAD

**Discussion by walterdeboer** [DependencyTrack/dependency-track#2559] [Show and tell]
Outdated Components is in the OWASP Top 10: A06:2021 – Vulnerable and Outdated Components

Dependency Track knows of outdated component, but does not show them in the Audit Vulnerabilities overview when there is no known vulnerability

If added two PR's so outdated component with no known vulnerabilities are included in the Audit Vulnerabilities overview :

- https://github.com/DependencyTrack/dependency-track/pull/2521
- https://github.com/DependencyTrack/frontend/pull/435

It uses the folowing PR so only stable versions are reported:
- https://github.com/DependencyTrack/dependency-track/pull/2501

N.B. only direct dependencies are reported as these are the ones you could update yourself

closes https://github.com/DependencyTrack/dependency-track/issues/2514, https://github.com/DependencyTrack/dependency-track/issues/2500, fixes https://github.com/DependencyTrack/dependency-track/issues/513, https://github.com/DependencyTrack/dependency-track/issues/1374, partly fixes https://github.com/DependencyTrack/dependency-track/issues/2003, https://github.com/DependencyTrack/dependency-track/issues/1833

![image](https://user-images.githubusercontent.com/11032733/222891065-f0cdd36d-b65f-4283-b946-2cea4d3b55ad.png)

Please share your thoughts! :-)

**Comment by stevespringett** [comment]
I would not expect to find outdated components while in the "Audit Vulnerabilities" tab.  Most large software applications with hundreds or thousands of OSS components will use a lot of out-of-date components, most will not be vulnerable. It will be frustrating for users to have to paginate through the noise to find the vulnerabilities. This would be a UX disaster if implemented.

This information should either be on a dedicated tab, or not included at all since users can create policy for outdated components which would then trigger a violation.

**Reply by walterdeboer** [reply]
Hi Steve, thanks for your reply. I've restricted the  outdated components to the direct components only, to prevent frustration for the users.  Only the components are shown, that you can update yourself. I think this makes it actionable and "Auditable"? Although I agree "Violations" does not cover it,

I'm not sure what you mean by "since users can create policy for outdated components"? Are you refering to #2537 ? Dependency Track now only supports a Component Age policy and a Component Version policy, but both are unsuitable to detect outdated components as you would need to know if there is a newer stable version available. Another issue with the policy overview is that it's missing the "evidence". I'd like to see the latest version available.

I hope you do see the value added by reporting the outdated components and showing the latest stable version available? I'm open to moving the outdated components on a dedicated tab (and a separate endpoint). The new tab could show:
- an overview of outdated components
- the latest stable version for each outdated component
- current vulnerabilities if any for each outdated component
- solved vulnerabilities in the latest version? This would probably need some extra effort.

Please advice on how to proceed?

**Reply by walterdeboer** [reply]
Or maybe just an extra filter switch on the components tab to exclude al non-outdated components would do the same?

![image](https://user-images.githubusercontent.com/11032733/222956189-7994c5f9-65a7-4770-bf4c-380c6cd303ff.png)

**Reply by stevespringett** [reply]
Hmmm. The outdated only switch is really interesting. I like that approach.  I could see this morphing into a dropdown with a few more options in the future, such as listing only the components whos hash value does not match the repository, etc.

**Reply by nscuro** [reply]
> Another issue with the policy overview is that it's missing the "evidence". I'd like to see the latest version available.

That is a more general enhancement that I also thought about. It would be great if we had a human-readable justification for each policy condition violation, similar to assertion failure messages produced by Hamcrest / AssertJ. e.g.

```
Expected component license "AGPL-3.0" to not be contained in license group "Copyleft", but it is
```

or

```
Expected the maximum distance to newest available version to be major=2/minor=/bugfix=, but was major=3/minor=1/bugfix=0
```

That would provide more context to the user as to what was expected, and why exactly that expectation was violated.

**Reply by walterdeboer** [reply]
Ok, a switch it will be :-)

https://github.com/DependencyTrack/dependency-track/pull/2568
https://github.com/DependencyTrack/frontend/pull/437

![image](https://user-images.githubusercontent.com/11032733/223234513-57f1664a-f8b3-44ec-8343-21e5b1712a91.png)
